lib/Dialect/TritonGPU/Transforms/Pipeliner/PipelineExpander.cpp

```diff
@@ -119,24 +119,38 @@ bool LoopPipelinerInternal::initializeLoopInfo(
   step = forOp.getStep();
 
   dynamicLoop = true;
-  auto upperBoundCst = ub.getDefiningOp<arith::ConstantIndexOp>();
-  auto lowerBoundCst = lb.getDefiningOp<arith::ConstantIndexOp>();
-  auto stepCst = step.getDefiningOp<arith::ConstantIndexOp>();
+
+  auto getIntegerAttrFromValue =
+      [](mlir::Value value) -> std::optional<mlir::IntegerAttr> {
+    auto constantOp = value.getDefiningOp<arith::ConstantOp>();
+    if (!constantOp) {
+      // Not defined by a ConstantOp
+      return std::nullopt;
+    }
+    // return constantOp.getValue().dyn_cast<mlir::IntegerAttr>();
+    return mlir::cast<mlir::IntegerAttr>(constantOp.getValue());
+  };
+  auto upperBoundCst = getIntegerAttrFromValue(ub);
+  auto lowerBoundCst = getIntegerAttrFromValue(lb);
+  auto stepCst = getIntegerAttrFromValue(step);
   if (!upperBoundCst || !lowerBoundCst || !stepCst) {
     if (!options.supportDynamicLoops) {
       LDBG("--dynamic loop not supported -> BAIL");
       return false;
     }
   } else {
-    int64_t ubImm = upperBoundCst.value();
-    int64_t lbImm = lowerBoundCst.value();
-    int64_t stepImm = stepCst.value();
+    int64_t ubImm = upperBoundCst->getInt();
+    int64_t lbImm = lowerBoundCst->getInt();
+    int64_t stepImm = stepCst->getInt();
     int64_t numIteration = llvm::divideCeilSigned(ubImm - lbImm, stepImm);
-    if (numIteration > maxStage) {
+    if (numIteration > options.numStages - 1) {
```

Comment:
In what case `maxStage` is different than `options.numStages - 1`?

Reply:
At this line, maxStage is 0, which is different from options.numStages - 1. This change was made per [Pawel's comment](https://github.com/triton-lang/triton/pull/4689#discussion_r1777657379).

Reply:
so if maxStage is 0 numStages is 1. Pawel's comment explains it: `maxStage == numStages-1`

Reply:
I see. In that case, did this line check if there's at least one iteration? Do we still need to check the number of iterations larger than num_stages from tt level?

Reply:
so if there is 0 iterations this condition is always false. I'm not sure I understand the questions

Reply:
Ah sorry for the confusion.
1) I saw `numStages` in `SoftwarePipeliner.cpp` was from the for loop to pipeline. In my PR, numStages was made accessible in `PipelineExpander.cpp`.  
2) I actually don't know how `maxStage == numStages-1` is preserved in the algorithm, but at this line, `maxStage` is zero, which is different from `numStages - 1`.
3) I believe the code around this line is about checking whether the number of loop iterations is smaller than the number of stages, which decides how the loop should be pipelined.
4) In the rest of `PipelineExpander.cpp`, maxStage is only updated [here](https://github.com/triton-lang/triton/blob/main/lib/Dialect/TritonGPU/Transforms/Pipeliner/PipelineExpander.cpp#L157). During the algorithm, it seems `maxStage` is gradually updated to a larger value than 0, while `maxStage <= numStages - 1` is satisfied throughout the algorithm. 
5) I left `maxStage == 0` unchanged to make sure the above algorithm in 4) is still the same. So, here in this line, I use `options.numStages - 1` to check the condition in 3).

Reply:
ah right, we should move the calculation of maxStage up

Reply:
Yeah the problem is that the actual maxStage depends on the pipeline schedule and here we are trying to early exit before getting a schedule, when we know the iteration count is too small for pipelining to be beneficial. So we are using numStages - 1 to check for early exit here.

Reply:
Do I need further improvement on this PR?

Reply:
This PR looks okay to me. Is it possible to upload part of patch to upstream mlir?